### PR TITLE
🚧Improve render system

### LINKED ISSUE #451: 🚧Improve render system
### Complete The Item Below

- [X] I have updated the title without removing the 🚧 emoji.

### Description

Currently, **_Velaptor_** has a class named `RendererFactory`.  The idea of this factory was to make it easy to create a renderer for the use of rendering whatever it is that you want to render.   It also currently only has a single method named `CreateRenderer()`.  This single method creates an `IRenderer` object.   The purpose of this is to allow a user to create the renderer (which is a singleton) and then use it to render whatever they want.

The problem is that the current implementation of `IRenderer` which is an `internal` class is getting too large and difficult to maintain, manage, and test.  This can be greatly improved upon.
The idea is to create different `IRender` types.  Each one will be specialized for rendering different kinds of things.  For example, one could be for rendering textures, one for line primitives, and another for text, etc.

**Things To Do:**

1. Create a new render texture overload to make it easier to render atlas sub textures
    - Create window size reactable event
    - Update window size throughout the codebase where it is required
    - Convert AtlasSubTextureData to struct
    - Inside atlas data, organize the sub-texture data into groups by sub texture id
      This is going to be called inside of the texture renderer, so it needs to be performant for `IAtlasData.GetFrames()`
    - Create a new render texture overload that takes an IAtlasData for rendering

2. Create 5 different types of render interfaces and associated implementations. 
The idea is that the `RendererFactory` class can be used by a user to easily generate any of these renderer types as needed.

All of the current unit tests of course will need to be migrated to the correct implementations.  

**Benefits:**
- More control is given to the user.
- Ability to only use what is required.
- Greatly cleans up the code base
- Greatly improves testability
- Reduces changes for bugs.  There are more rendering capabilities coming in the future and this will help create a better path to do so.

**Things To Consider:**
> **Note**
> All of the renderer implementations need to be `internal sealed` so the users do not have access.

> **Note**
> All of the renderer implementations need to go into the `Velaptor.Graphics` namespace

> **Note**
> The `IBatchServiceManager` and `BatchServiceManager` types might not be needed anymore.  This is because
> they will not all be required in a single place like it is currently set up in the `Renderer` class.  Each renderer can simply pull in
> the required batch manager that is required via DI.

> **Note**
> The _**Velaptor Testing**_ application scene management system needs to be updated.  This is going to eventually become a major feature in **_Velaptor_** itself instead of a **_Velaptor Testing_** only feature.  The `IRenderer` needs to be removed
from the scene management system by removing the `IRenderer` parameter from the `Render()` method and replaced with a `FrameTime` parameter.

**List Of Render Types:**
- `ITextureRenderer`
- `IFontRenderer`
- `IRectangleRenderer`
- `ILineRenderer`

> **https://github.com/KinsonDigital/Velaptor/labels/%F0%9F%A7%A8breaking%20changes**
- `IRenderer` removed
- `RenderFactory.CreateRenderer()` has been removed
- `RenderFactory` has been changed to sealed

### Acceptance Criteria

- [x] Create window size reactable event
    - This means a new subscription in the IRenderer implementation to receive the size of the window
- [x] Remove the `IAtlasData.GetFrame()` method - https://github.com/KinsonDigital/Velaptor/labels/%F0%9F%A7%A8breaking%20changes
- [x] Refactor code that uses the term _**Atlas**_ - https://github.com/KinsonDigital/Velaptor/labels/%F0%9F%A7%A8breaking%20changes
- [x] Improve the `IAtlasData.GetFrames` method
    - This involves grouping the data in the beginning in the `ctor`
- [x] Refactor code everywhere where the window size is required.
- [x] `ITextureRenderer`
    - [x] Create implementation
    - [x] Create unit tests
    - [x] Class is `internal sealed`
    - [x] Create new overloads for easy sprite sheet rendering using the signature below:
      - This will be base call that all other texture atlas calls will end up invoking
      ```cs
        private void RenderSubTexture(
            ITexture texture,
            IAtlasData atlasData,
            string subTextureId,
            uint frame,
            int x,
            int y,
            float size,
            float angle,
            Color color,
            RenderEffects effects,
            int layer = 0)
        {
        }
      ```
- [x] `IFontRenderer`
    - [x] Create implementation
    - [x] Create unit tests
    - [x] Class is `internal sealed`
- [x] `IRectangleRenderer`
    - [x] Create implementation
    - [x] Create unit tests
    - [x] Class is `internal sealed`
- [x] `ILineRenderer`
    - [x] Create implementation
    - [x] Create unit tests
    - [x] Class is `internal sealed`
- [x] `RendererFactory` changes
    - [x] Create a new method named `CreateTextureRenderer()`
    - [x] Create a new method named `CreateFontRenderer()`
    - [x] Create a new method named `CreateRectangleRenderer()`
    - [x] Create a new method named `CreateLineRenderer()`
- [x] `IRenderer` interface deleted and cleaned up
- [x] `Renderer` class deleted and cleaned up

### ToDo Items

- [X] Change type labels added to this issue.  Refer to the _**Change Type Labels**_ section below.
- [X] Priority label added to this issue.  Refer to the _**Priority Type Labels**_ section below.
- [X] Issue linked to the correct project _(if applicable)_.
- [X] Issue linked to the correct milestone _(if applicable)_.
- [x] Draft pull request created and linked to this issue _(only required with code changes)_.

### Issue Dependencies

_No response_

### Related Work

_No response_

### Additional Information:

**_<details closed><summary>Change Type Labels</summary>_**

| Change Type         | Label                |
|---------------------|----------------------|
| Bug Fixes           | `🐛bug` |
| Breaking Changes    | `🧨breaking changes` |
| New Feature         | `✨new feature` |
| Workflow Changes    | `workflow` |
| Code Doc Changes    | `🗒️documentation/code` |
| Product Doc Changes | `📝documentation/product` |
</details>

**_<details closed><summary>Priority Type Labels</summary>_**

| Priority Type       | Label             |
|---------------------|-------------------|
| Low Priority        | `low priority`    |
| Medium Priority     | `medium priority` |
| High Priority       | `high priority`   |
</details>

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct.

### PR DESCRIPTION
**Pull request by CalvinWilkinson**
<h1 style="font-weight:bold" align="center">Preview Feature Pull Request</h1>
<h2 style="font-weight:bold" align="center">📋Review Checklist📋</h2>

- [x] The **_head(source)_** branch for this pull request is a **_preview feature_** branch, with the correct naming syntax, in the following example:
  - A **_head(source)_** branch is the branch being merged into a target branch.
    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/feature/<issue-num>-<description>
      Example: preview/feature/123-my-preview-feature
      ```
    </details>

- [x] The **_head(source)_** branch for this pull request is created from a **_preview release_** branch, with the correct naming syntax, in the following example:
  <details closed><summary>Example</summary>

    ``` xml
    Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
    Example: preview/v1.2.3-preview.4
    ```
  </details>

- [x] The **_base(target)_** branch for this pull request is a **_preview release_** branch, with the correct naming syntax, in the following example:
  - A **_base(target)_** branch is the branch that the **_head(source)_** branch is merging into.

    <details closed><summary>Example</summary>

      ``` xml
      Syntax: preview/v<major>.<minor>.<patch>-preview.<prev-num>
      Example: preview/v1.2.3-preview.4
      ```
    </details>

💡For more information on branching, refer to the project's [branching documentation](../../Documentation/Branching.md).

- [x] Associated issue exists and is linked to this pull request.
  - One issue per pull request.

- [x] Pull request title matches the title of the associated issue.

- [x] A `preview` label has been added to the pull request.

- [x] The labels attached to this PR match the labels attached to the associated issue.

- [x] Project assigned matches the project of the associated issue.

- [x] Milestone assigned matches the milestone of the associated issue.

- [x] My code follows the coding style of this project.
  - The style is enforced by the **_.editorconfig_** files in the project and displayed as warnings.  If there is an edge case, reach out, and let's discuss it.

- [x] I have manually tested my code changes to the best of my ability.

- [x] All tests passed locally.
  - This is required because unit tests might pass locally but not in the CI environment during the status check process or vice-versa.
  - Tests might pass on the developer's machine but not necessarily on the code reviewer's machine.
  - Status checks run for this pull request when merging into the **_preview/v\*.\*.\*-preview.\*_** branch.  These status checks run every time a change is pushed to the pull request.  These checks validate version syntax, tagging, builds, unit tests, and more.

### CODE REVIEW

File: Testing/VelaptorTests/OpenGL/Buffers/FontGPUBufferTests.cs [hunk 1]
```diff
@@ -60,21 +62,64 @@ public FontGPUBufferTests()
...
         this.mockReactable = new Mock<IReactable>();
         this.mockReactable.Setup(m => m.Subscribe(It.IsAny<IReactor>()))
-            .Callback<IReactor>(reactor =>
+            .Returns<IReactor>(reactor =>
             {
                 reactor.Should().NotBeNull("it is required for unit testing.");
 
+                if (reactor.EventId == NotificationIds.GLInitializedId)
+                {
+                    // RETURN NULL TO SIMPLY IGNORE THIS EVENT ID
```
Comment (kselena) [inline Testing/VelaptorTests/OpenGL/Buffers/FontGPUBufferTests.cs:76]: Refactor to remove simply
```suggestion
                    // RETURN NULL TO IGNORE THIS EVENT ID
```

File: Testing/VelaptorTests/OpenGL/Buffers/RectGPUBufferTests.cs [hunk 2]
```diff
@@ -63,21 +65,64 @@ public RectGPUBufferTests()
...
         this.mockReactable = new Mock<IReactable>();
         this.mockReactable.Setup(m => m.Subscribe(It.IsAny<IReactor>()))
-            .Callback<IReactor>(reactor =>
+            .Returns<IReactor>(reactor =>
             {
                 reactor.Should().NotBeNull("it is required for unit testing.");
 
+                if (reactor.EventId == NotificationIds.GLInitializedId)
+                {
+                    // RETURN NULL TO SIMPLY IGNORE THIS EVENT ID
```
Comment (kselena) [inline Testing/VelaptorTests/OpenGL/Buffers/RectGPUBufferTests.cs:79]: Refactor to remove simply
```suggestion
                    // RETURN NULL TO IGNORE THIS EVENT ID
```

File: Velaptor/Graphics/IRenderer.cs [hunk 3]
```diff
@@ -17,18 +17,6 @@ namespace Velaptor.Graphics;
...
-    uint RenderSurfaceWidth { get; set; }
-
-    /// <summary>
-    /// Gets or sets the render surface height.
-    /// </summary>
-    /// <remarks>This is the height of the viewport.</remarks>
-    uint RenderSurfaceHeight { get; set; }
-
     /// <summary>
     /// Gets or sets the color that the back buffer will be cleared to.
```
Comment (kselena) [inline Velaptor/Graphics/IRenderer.cs]: Grammar
```suggestion
    /// Gets or sets the color where the back buffer will be cleared.
```

File: Velaptor/OpenGL/Buffers/TextureGPUBuffer.cs [hunk 4]
```diff
@@ -125,10 +125,10 @@ protected internal override void UploadVertexData(TextureBatchItem textureQuad,
...
-        var vertex1 = topLeft.ToNDC(textureQuad.ViewPortSize.Width, textureQuad.ViewPortSize.Height);
-        var vertex2 = bottomLeft.ToNDC(textureQuad.ViewPortSize.Width, textureQuad.ViewPortSize.Height);
-        var vertex3 = topRight.ToNDC(textureQuad.ViewPortSize.Width, textureQuad.ViewPortSize.Height);
-        var vertex4 = bottomRight.ToNDC(textureQuad.ViewPortSize.Width, textureQuad.ViewPortSize.Height);
+        var vertex1 = topLeft.ToNDC(this.ViewPortSize.Width, this.ViewPortSize.Height);
+        var vertex2 = bottomLeft.ToNDC(this.ViewPortSize.Width, this.ViewPortSize.Height);
+        var vertex3 = topRight.ToNDC(this.ViewPortSize.Width, this.ViewPortSize.Height);
+        var vertex4 = bottomRight.ToNDC(this.ViewPortSize.Width, this.ViewPortSize.Height);
 
         // Set up the corners of the sub texture to render
```
Comment (kselena) [inline Velaptor/OpenGL/Buffers/TextureGPUBuffer.cs:135]: Grammar
```suggestion
        // Sets the corners of the sub texture to render
```

File: Velaptor/OpenGL/GLWindow.cs [hunk 5]
```diff
@@ -393,15 +393,15 @@ private void GLWindow_Closing()
     /// </summary>
     private void GLWindow_Resize(Vector2D<int> obj)
     {
-        var uWidth = (uint)obj.X;
-        var uHeight = (uint)obj.Y;
+        var width = (uint)obj.X;
+        var height = (uint)obj.Y;
 
         // Update the viewport so it is the same size as the window
```
Comment (kselena) [inline Velaptor/OpenGL/GLWindow.cs:394]: Grammar
```suggestion
        // Updates the viewport so it is the same size as the window
```

File: Velaptor/OpenGL/TextureBatchItem.cs [hunk 6]
```diff
@@ -25,7 +25,6 @@ namespace Velaptor.OpenGL;
     /// <param name="angle">The angle of the texture in degrees.</param>
     /// <param name="tintColor">The color to apply to an entire texture.</param>
```
Comment (kselena) [inline Velaptor/OpenGL/TextureBatchItem.cs:26]: Wordy
```suggestion
    /// <param name="tintColor">The color to apply to a texture.</param>
```

File: Velaptor/ReactableData/ViewPortSizeData.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+﻿// <copyright file="ViewPortSizeData.cs" company="KinsonDigital">
+// Copyright (c) KinsonDigital. All rights reserved.
+// </copyright>
+
+namespace Velaptor.ReactableData;
+
+/// <summary>
+/// Holds the size of the view port.
```
Comment (kselena) [inline Velaptor/ReactableData/ViewPortSizeData.cs:8]: Refactor - should be one word
```suggestion
/// Holds the size of the viewport.
```

**Review by kselena — CHANGES_REQUESTED** [review]

**Review by KinsonDigitalAdmin — DISMISSED** [review]

**Review by KinsonDigitalAdmin — APPROVED** [review]

**Review by kselena — APPROVED** [review]

**Pull request merged**